microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/internal/metrics/properties/MicroProfileMetricsProvider.java

```diff
@@ -0,0 +1,75 @@
+/*******************************************************************************
+* Copyright (c) 2020 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package com.redhat.microprofile.jdt.internal.metrics.properties;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.IAnnotation;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.search.SearchMatch;
+
+import com.redhat.microprofile.jdt.core.AbstractAnnotationTypeReferencePropertiesProvider;
+import com.redhat.microprofile.jdt.core.IPropertiesCollector;
+import com.redhat.microprofile.jdt.core.SearchContext;
+import com.redhat.microprofile.jdt.core.utils.JDTTypeUtils;
+
+/**
+ * Properties provider that provides MicroProfile Metrics properties from the
+ * 
+ * @author David Kwon
+ * 
+ * @see https://github.com/eclipse/microprofile-metrics/blob/dc94b84ec90dd4a0bc983336e4273247a4e415cc/spec/src/main/asciidoc/architecture.adoc
+ *
+ */
+public class MicroProfileMetricsProvider extends AbstractAnnotationTypeReferencePropertiesProvider {
```

Comment:
I think we should simplify provider class which wishes to implement static properties. Here my suggestion:

 * create abstract class AbstractStaticPropertiesProvider like this (please change name of class, method if you don't like it and please comment each methods):

```java
package com.redhat.microprofile.jdt.core;

import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.jdt.core.search.SearchMatch;
import org.eclipse.jdt.core.search.SearchPattern;

public abstract class AbstractStaticPropertiesProvider extends AbstractPropertiesProvider {

	@Override
	public final void beginSearch(SearchContext context, IProgressMonitor monitor) {
		if (isAdaptedFor(context, monitor)) {
			collectStaticProperties(context, monitor);
		}
	}

	protected abstract boolean isAdaptedFor(SearchContext context, IProgressMonitor monitor);

	protected abstract void collectStaticProperties(SearchContext context, IProgressMonitor monitor);

	@Override
	public void collectProperties(SearchMatch match, SearchContext context, IProgressMonitor monitor) {
		// Do nothing
	}

	@Override
	protected String[] getPatterns() {
		return null;
	}

	@Override
	protected SearchPattern createSearchPattern(String pattern) {
		return null;
	}

}

```

 * implements this class like this:

```java
package com.redhat.microprofile.jdt.core;

import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.jdt.core.IJavaProject;

import com.redhat.microprofile.jdt.core.utils.JDTTypeUtils;

public class MicroProfileMetricsProvider extends AbstractStaticPropertiesProvider{

	@Override
	protected boolean isAdaptedFor(SearchContext context, IProgressMonitor monitor) {
		IJavaProject javaProject = context.getJavaProject();
		return (JDTTypeUtils.findType(javaProject, METRIC_ID) == null);
	}

	@Override
	protected void collectStaticProperties(SearchContext context, IProgressMonitor monitor) {
		IPropertiesCollector collector = context.getCollector();
		String docs = "List of tag values.\r\n"
				+ "Tag values set through `mp.metrics.tags` MUST escape equal symbols `=` and commas `,` with a backslash `\\`.";
		super.addItemMetadata(collector, GLOBAL_TAGS_VARIABLE,
				"java.util.Optional<java.lang.String>", docs, null, null, null, null, null, false);

		docs = "The app name.";
		super.addItemMetadata(collector, APPLICATION_NAME_VARIABLE,
				"java.util.Optional<java.lang.String>", docs, null, null, null, null, null, false);
	}

}

```

Reply:
This looks great!

In my latest  push, for the `AbstractStaticPropertiesProvider#getPatterns` method, I decided to return an empty array rather than null, since returning null created an NPE in `AbstractPropertiesProvider#createSearchPattern` here:

```
...
String[] patterns = getPatterns();
    for (String pattern : patterns) {
    ...
```

Reply:
> This looks great!

Glad you like it.

>  I decided to return an empty array rather than null, since returning null created an NPE

I think it should be better to return null and fix the NPE in AbstractPropertiesProvider#createSearchPattern

Reply:
Thanks, fixed